// spectacle/es/hooks/use-slide.js
import React from 'react';
import { DeckContext } from './use-deck';
import { DEFAULT_SLIDE_ELEMENT_INDEX } from '../utils/constants';
/**
 * Performs logic operations for all of the slide domain level.
 * slideElements are dynamic elements within the slide that may
 * appear/ move around etc.
 * If remaining elements in slide, these are brought in one by one.
 * If not, we tell the deck to take us to the next slide.
 *
 * Note: Immediate is a React-Spring property that we pass to the animations
 * essentially it skips animations.
 */
// Initialise SlideContext.

export var SlideContext = React.createContext();

function useSlide(slideNum) {
  // Gets state, dispatch and number of slides off DeckContext.
  var _React$useContext = React.useContext(DeckContext),
      deckContextState = _React$useContext.state,
      deckContextDispatch = _React$useContext.dispatch,
      slideElementMap = _React$useContext.slideElementMap,
      numberOfSlides = _React$useContext.numberOfSlides;

  if (slideNum === 'undefined') {
    throw new Error('Must provide slide number to useSlide. Provided undefined instead.');
  }

  var reverseDirection = deckContextState.reverseDirection,
      immediate = deckContextState.immediate;
  var slideElementsLength = slideElementMap[slideNum];
  var isActiveSlide = deckContextState.currentSlide === slideNum;
  var setNotes = React.useCallback(function (notes) {
    deckContextDispatch({
      type: 'SET_NOTES',
      payload: {
        notes: notes,
        slideNumber: slideNum
      }
    });
  }, [deckContextDispatch, slideNum]);
  var currentSlideElement = isActiveSlide ? deckContextState.currentSlideElement : DEFAULT_SLIDE_ELEMENT_INDEX;
  return {
    state: {
      reverseDirection: reverseDirection,
      slideElementsLength: slideElementsLength,
      currentSlideElement: currentSlideElement,
      immediate: immediate,
      isActiveSlide: isActiveSlide,
      numberOfSlides: numberOfSlides
    },
    actions: {
      setNotes: setNotes
    }
  };
}

export default useSlide;
