// spectacle/es/components/code-pane.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import Highlight, { defaultProps } from 'prism-react-renderer';
import propTypes from 'prop-types';
import theme from 'prism-react-renderer/themes/vsDark';
import lightTheme from 'prism-react-renderer/themes/nightOwlLight';
import { ThemeContext } from 'styled-components';
import { DeckContext } from '../hooks/use-deck';
var spaceSearch = /\S|$/;
var lineNumberStyles = {
  padding: '0 1em',
  borderRight: '1px solid hsla(0, 0%, 100%, 0.25)',
  flex: '0 1 30px',
  alignSelf: 'stretch'
};
export default function CodePane(props) {
  var canvas = React.useRef(document.createElement('canvas'));
  var context = React.useRef(canvas.current.getContext('2d'));
  var themeContext = React.useContext(ThemeContext);

  var _React$useContext = React.useContext(DeckContext),
      printMode = _React$useContext.state.printMode;

  var font = React.useMemo(function () {
    if (themeContext && themeContext.fonts && themeContext.fonts.monospace) {
      return themeContext.fonts.monospace;
    }

    var _navigator = navigator,
        platform = _navigator.platform;

    if (platform.toLowerCase().search('win') !== -1) {
      return 'Consolas';
    } else if (platform.toLowerCase().search('mac') !== -1) {
      return 'Menlo';
    } else {
      return 'monospace';
    }
  }, [themeContext]);
  var fontSize = React.useMemo(function () {
    if (themeContext && themeContext.fontSizes && themeContext.fontSizes.monospace) {
      return themeContext.fontSizes.monospace;
    }

    return props.fontSize;
  }, [themeContext, props.fontSize]);
  var preStyles = React.useMemo(function () {
    return {
      fontFamily: font,
      fontSize: fontSize,
      maxHeight: themeContext.size.maxCodePaneHeight || 200,
      overflow: 'scroll',
      margin: 0,
      padding: '0.5em 1em 0.5em 0'
    };
  }, [font, fontSize, themeContext]);
  var measureIndentation = React.useCallback(function (indentation) {
    if (indentation === 0) {
      return 0;
    }

    var string = ' '.repeat(indentation);
    context.current.font = "".concat(props.fontSize, "px ").concat(font);
    var measurement = context.current.measureText(string);
    return measurement.width;
  }, [props.fontSize, font]);
  return React.createElement(React.Fragment, null, React.createElement(Highlight, _extends({}, defaultProps, {
    code: props.children,
    language: props.language,
    theme: printMode ? lightTheme : theme
  }), function (_ref) {
    var className = _ref.className,
        style = _ref.style,
        tokens = _ref.tokens,
        getLineProps = _ref.getLineProps,
        getTokenProps = _ref.getTokenProps;
    return React.createElement("pre", {
      className: "".concat(className, " ").concat(props.autoFillHeight && 'spectacle-auto-height-fill'),
      style: _objectSpread({}, style, {}, preStyles)
    }, tokens.map(function (line, i) {
      var lineProps = getLineProps({
        line: line,
        key: i
      });
      var lineIndentation = line[0].content.search(spaceSearch);
      lineProps.style = _objectSpread({}, lineProps.style || {}, {
        whiteSpace: 'pre-wrap',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'flex-start'
      });

      if (line[0].content && !line[0].empty) {
        line[0].content = line[0].content.trimLeft();
      }

      return React.createElement("div", _extends({
        key: i
      }, lineProps), React.createElement("div", {
        style: lineNumberStyles
      }, i + 1), React.createElement("div", {
        style: {
          marginLeft: measureIndentation(lineIndentation),
          flex: 1,
          paddingLeft: '0.25em'
        }
      }, line.map(function (token, key) {
        return React.createElement("span", _extends({
          key: key
        }, getTokenProps({
          token: token,
          key: key
        })));
      })));
    }));
  }));
}
CodePane.propTypes = {
  autoFillHeight: propTypes.bool,
  children: propTypes.string.isRequired,
  fontSize: propTypes.number,
  language: propTypes.string.isRequired,
  theme: propTypes.object
};
CodePane.defaultProps = {
  language: 'javascript',
  theme: theme,
  fontSize: 15
};
