// wonka/lib/js/src/web/wonkaJs.js
'use strict';

var Wonka_callbag = require("./wonka_callbag.js");
var Wonka_observable = require("./wonka_observable.js");
var Wonka_operator_delay = require("./wonka_operator_delay.js");
var Wonka_sink_toPromise = require("./wonka_sink_toPromise.js");
var Wonka_source_interval = require("./wonka_source_interval.js");
var Wonka_operator_debounce = require("./wonka_operator_debounce.js");
var Wonka_operator_throttle = require("./wonka_operator_throttle.js");
var Wonka_source_fromPromise = require("./wonka_source_fromPromise.js");
var Wonka_source_fromDomEvent = require("./wonka_source_fromDomEvent.js");
var Wonka_source_fromListener = require("./wonka_source_fromListener.js");

var debounce = Wonka_operator_debounce.debounce;

var delay = Wonka_operator_delay.delay;

var throttle = Wonka_operator_throttle.throttle;

var toPromise = Wonka_sink_toPromise.toPromise;

var interval = Wonka_source_interval.interval;

var fromDomEvent = Wonka_source_fromDomEvent.fromDomEvent;

var fromListener = Wonka_source_fromListener.fromListener;

var fromPromise = Wonka_source_fromPromise.fromPromise;

var fromObservable = Wonka_observable.fromObservable;

var toObservable = Wonka_observable.toObservable;

var fromCallbag = Wonka_callbag.fromCallbag;

var toCallbag = Wonka_callbag.toCallbag;

exports.debounce = debounce;
exports.delay = delay;
exports.throttle = throttle;
exports.toPromise = toPromise;
exports.interval = interval;
exports.fromDomEvent = fromDomEvent;
exports.fromListener = fromListener;
exports.fromPromise = fromPromise;
exports.fromObservable = fromObservable;
exports.toObservable = toObservable;
exports.fromCallbag = fromCallbag;
exports.toCallbag = toCallbag;
/* Wonka_observable Not a pure module */
