// spectacle/es/components/deck/index.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  height: 100vh;\n  width: 100vw;\n  position: fixed;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeContext, ThemeProvider } from 'styled-components';
import normalize from 'normalize-newline';
import indentNormalizer from '../../utils/indent-normalizer';
import useDeck, { DeckContext } from '../../hooks/use-deck';
import isComponentType from '../../utils/is-component-type';
import useUrlRouting from '../../hooks/use-url-routing';
import PresenterDeck from './presenter-deck';
import AudienceDeck from './audience-deck';
import { mergeTheme } from '../../theme';
import { PrintDeck } from './print-deck';
import { animated, useTransition } from 'react-spring';
import { TransitionPipeContext, TransitionPipeProvider } from '../../hooks/use-transition-pipe';
import usePresentation, { MSG_SLIDE_STATE_CHANGE } from '../../hooks/use-presentation';
import useKeyboardControls from '../../hooks/use-keyboard-controls';
import { DEFAULT_SLIDE_ELEMENT_INDEX, DEFAULT_SLIDE_INDEX } from '../../utils/constants';
import searchChildrenForAppear from '../../utils/search-children-appear';
import OverviewDeck from './overview-deck';
import { Markdown, Slide, Notes } from '../../index';
import { isolateNotes, removeNotes } from '../../utils/notes';
var AnimatedDeckDiv = styled(animated.div)(_templateObject());
var defaultSlideEffect = {
  from: {
    position: 'fixed',
    transform: 'translate(100%, 0%)'
  },
  enter: {
    position: 'fixed',
    transform: 'translate(0, 0%)'
  },
  leave: {
    position: 'fixed',
    transform: 'translate(-100%, 0%)'
  },
  config: {
    precision: 0
  }
};
/**
 * Provides top level state/context provider with useDeck hook
 * Should wrap all the presentation components (slides, etc)
 *
 * Props = {
 *  loop: bool (pass in true if you want slides to loop)
 * transitionEffect: based off of react sprint useTransition
 * }
 *
 * Note: Immediate is a React-Spring property that we pass to the animations
 * essentially it skips animations.
 */

var initialState = {
  currentSlide: DEFAULT_SLIDE_INDEX,
  immediate: false,
  immediateElement: false,
  currentSlideElement: DEFAULT_SLIDE_ELEMENT_INDEX,
  reverseDirection: false,
  presenterMode: false,
  overviewMode: false,
  notes: {},
  resolvedInitialUrl: false
};

var mapMarkdownIntoSlides = function mapMarkdownIntoSlides(child, index) {
  if (isComponentType(child, Markdown.name) && Boolean(child.props.containsSlides)) {
    return child.props.children.split(/\n\s*---\n/).map(function (markdown, mdIndex) {
      var content = normalize(indentNormalizer(markdown));
      var contentWithoutNotes = removeNotes(content);
      var notes = isolateNotes(content);
      return React.createElement(Slide, {
        key: "md-slide-".concat(index, "-").concat(mdIndex)
      }, React.createElement(Markdown, null, contentWithoutNotes), React.createElement(Notes, null, notes));
    });
  }

  return child;
};

var Deck = function Deck(_ref) {
  var children = _ref.children,
      loop = _ref.loop,
      keyboardControls = _ref.keyboardControls,
      animationsWhenGoingBack = _ref.animationsWhenGoingBack,
      rest = _objectWithoutProperties(_ref, ["children", "loop", "keyboardControls", "animationsWhenGoingBack"]);

  if (React.Children.count(children) === 0) {
    throw new Error('Spectacle must have at least one slide to run.');
  }

  var filteredChildren = React.Children.map(children, mapMarkdownIntoSlides).reduce(function (acc, slide) {
    return acc.concat(slide);
  }, []).filter(function (child) {
    return isComponentType(child, Slide.name);
  });
  var numberOfSlides = filteredChildren.length;

  if (numberOfSlides === 0) {
    throw new Error('Spectacle must have at least one slide to run.');
  }

  var slideElementMap = React.useMemo(function () {
    var map = {};
    filteredChildren.filter(function (slide, index) {
      map[index] = searchChildrenForAppear(slide.props.children);
    });
    return map;
  }, [filteredChildren]); // Initialise useDeck hook and get state and dispatch off of it

  var _useDeck = useDeck(_objectSpread({}, initialState, {
    numberOfSlides: numberOfSlides
  })),
      state = _useDeck.state,
      dispatch = _useDeck.dispatch;

  var themeContext = React.useContext(ThemeContext);
  React.useLayoutEffect(function () {
    document.body.style.margin = '0';
    document.body.style.background = '#000';
    document.body.style.color = themeContext.colors[rest.textColor] || rest.textColor || themeContext.colors.primary;
  }, [rest.backgroundColor, rest.textColor, themeContext.colors]);

  var _usePresentation = usePresentation(),
      startConnection = _usePresentation.startConnection,
      terminateConnection = _usePresentation.terminateConnection,
      sendMessage = _usePresentation.sendMessage,
      errors = _usePresentation.errors,
      addMessageHandler = _usePresentation.addMessageHandler,
      isReceiver = _usePresentation.isReceiver,
      isController = _usePresentation.isController;

  React.useEffect(function () {
    if (errors && errors.length > 0) {
      console.log('presentation errors', errors);
    }
  }, [errors]);
  var onUrlChange = React.useCallback(function (update) {
    if (isController) {
      sendMessage({
        type: MSG_SLIDE_STATE_CHANGE,
        payload: update
      });
    }
  }, [sendMessage, isController]);

  var _useUrlRouting = useUrlRouting({
    dispatch: dispatch,
    currentSlide: state.currentSlide,
    currentSlideElement: state.currentSlideElement,
    currentPresenterMode: state.presenterMode,
    slideElementMap: slideElementMap,
    loop: loop,
    animationsWhenGoingBack: animationsWhenGoingBack,
    onUrlChange: onUrlChange
  }),
      navigateToNext = _useUrlRouting.navigateToNext,
      navigateToPrevious = _useUrlRouting.navigateToPrevious,
      toggleMode = _useUrlRouting.toggleMode,
      goToSlide = _useUrlRouting.goToSlide;

  useKeyboardControls({
    keyboardControls: keyboardControls,
    navigateToNext: navigateToNext,
    navigateToPrevious: navigateToPrevious,
    toggleMode: toggleMode
  });

  var _React$useContext = React.useContext(TransitionPipeContext),
      runTransition = _React$useContext.runTransition;

  var userTransitionEffect = filteredChildren[state.currentSlide].props.transitionEffect || {};
  var transitionRef = React.useRef(null);
  React.useEffect(function () {
    if (!transitionRef.current) {
      return;
    }

    runTransition(transitionRef.current);
  }, [transitionRef, state.currentSlide, runTransition]);
  var transitions = useTransition(state.currentSlide, function (p) {
    return p;
  }, {
    ref: transitionRef,
    enter: function enter() {
      return userTransitionEffect.enter || defaultSlideEffect.enter;
    },
    leave: userTransitionEffect.leave || defaultSlideEffect.leave,
    from: userTransitionEffect.from || defaultSlideEffect.from,
    unique: true,
    immediate: state.immediate
  });
  var content = null;

  if (state.resolvedInitialUrl) {
    if (state.overviewMode) {
      var staticSlides = filteredChildren.map(function (slide, index) {
        return React.cloneElement(slide, {
          slideNum: index,
          template: rest.template
        });
      });
      content = React.createElement(OverviewDeck, {
        goToSlide: goToSlide
      }, staticSlides);
    } else if (state.exportMode) {
      var _staticSlides = filteredChildren.map(function (slide, index) {
        return React.cloneElement(slide, {
          slideNum: index,
          template: rest.template
        });
      });

      content = React.createElement(PrintDeck, null, _staticSlides);
    } else if (state.presenterMode) {
      var _staticSlides2 = filteredChildren.map(function (slide, index) {
        return React.cloneElement(slide, {
          slideNum: index,
          template: rest.template
        });
      });

      content = React.createElement(PresenterDeck, {
        isController: isController,
        isReceiver: isReceiver,
        startConnection: startConnection,
        terminateConnection: terminateConnection
      }, _staticSlides2);
    } else {
      var animatedSlides = transitions.map(function (_ref2) {
        var item = _ref2.item,
            props = _ref2.props,
            key = _ref2.key;
        return React.createElement(AnimatedDeckDiv, {
          style: props,
          key: key
        }, React.cloneElement(filteredChildren[item], {
          slideNum: item,
          numberOfSlides: numberOfSlides,
          template: rest.template
        }));
      });
      content = React.createElement(AudienceDeck, {
        addMessageHandler: addMessageHandler
      }, animatedSlides);
    }
  }

  return React.createElement(React.Fragment, null, React.createElement(DeckContext.Provider, {
    value: {
      state: state,
      dispatch: dispatch,
      numberOfSlides: numberOfSlides,
      keyboardControls: keyboardControls,
      animationsWhenGoingBack: animationsWhenGoingBack,
      slideElementMap: slideElementMap,
      goToSlide: goToSlide
    }
  }, content));
};

Deck.propTypes = {
  animationsWhenGoingBack: PropTypes.bool.isRequired,
  backgroundColor: PropTypes.string,
  children: PropTypes.node.isRequired,
  keyboardControls: PropTypes.oneOf(['arrows', 'space']),
  loop: PropTypes.bool.isRequired,
  template: PropTypes.func,
  textColor: PropTypes.string,
  theme: PropTypes.object
};

var ConnectedDeck = function ConnectedDeck(props) {
  return React.createElement(ThemeProvider, {
    theme: mergeTheme(props.theme)
  }, React.createElement(TransitionPipeProvider, null, React.createElement(Deck, props)));
};

ConnectedDeck.propTypes = Deck.propTypes;
ConnectedDeck.defaultProps = {
  loop: false,
  keyboardControls: 'arrows',
  animationsWhenGoingBack: false
};
export default ConnectedDeck;
