// spectacle/lib/components/table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableRow = exports.TableCell = exports.Table = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Table = (0, _styledComponents.default)('table')((0, _styledSystem.compose)(_styledSystem.color, _styledSystem.typography, _styledSystem.space, _styledSystem.border, _styledSystem.layout));
exports.Table = Table;
Table.defaultProps = {
  color: 'primary',
  fontFamily: 'text',
  fontSize: 'text',
  textAlign: 'left',
  margin: 'listMargin'
};
var TableRow = (0, _styledComponents.default)('tr')((0, _styledSystem.compose)(_styledSystem.color, _styledSystem.typography, _styledSystem.space, _styledSystem.border, _styledSystem.layout));
exports.TableRow = TableRow;
TableRow.defaultProps = {
  color: 'primary',
  fontFamily: 'text',
  fontSize: 'text',
  textAlign: 'left',
  margin: 'listMargin'
};
var TableCell = (0, _styledComponents.default)('td')((0, _styledSystem.compose)(_styledSystem.color, _styledSystem.typography, _styledSystem.space, _styledSystem.border, _styledSystem.layout));
exports.TableCell = TableCell;
TableCell.defaultProps = {
  color: 'primary',
  fontFamily: 'text',
  fontSize: 'text',
  textAlign: 'left',
  margin: 'listMargin'
};
