// wonka/lib/js/src/wonka_sources.js
'use strict';

var Wonka_source_make = require("./sources/wonka_source_make.js");
var Wonka_source_fromList = require("./sources/wonka_source_fromList.js");
var Wonka_source_fromArray = require("./sources/wonka_source_fromArray.js");
var Wonka_source_fromValue = require("./sources/wonka_source_fromValue.js");
var Wonka_source_primitives = require("./sources/wonka_source_primitives.js");
var Wonka_source_makeSubject = require("./sources/wonka_source_makeSubject.js");

var fromArray = Wonka_source_fromArray.fromArray;

var fromList = Wonka_source_fromList.fromList;

var fromValue = Wonka_source_fromValue.fromValue;

var make = Wonka_source_make.make;

var makeSubject = Wonka_source_makeSubject.makeSubject;

var empty = Wonka_source_primitives.empty;

var never = Wonka_source_primitives.never;

exports.fromArray = fromArray;
exports.fromList = fromList;
exports.fromValue = fromValue;
exports.make = make;
exports.makeSubject = makeSubject;
exports.empty = empty;
exports.never = never;
/* No side effect */
